Fix type casting exceptions in 'isof' and 'cast' calls.

WanjohiSammy opened the pull request:
<!-- markdownlint-disable MD002 MD041 -->

### Issues

*This pull request fixes [#1744](https://github.com/OData/odata.net/issues/1744).*

### Description

When using unquoted type parameters in `isof` and `cast`, the operations perform as expected provided that the source type can be cast to the target type. Issues arise when casting is not possible, resulting in the following exception:
```
Encountered invalid type cast. '{source type}' is not assignable from '{target type}'.
```
Lets say you have the following data model:
```cs
namespace NS;

public class Product
{
    public int ProductID { get; set; }
    // Others properties
    public Address SupplierAddress { get; set; }
}

public class DerivedProduct : Product
{
   public string AnotherProperty { get; set; }
    // Others properties
}

public class Address
{
    // Others properties
    public string City { get; set; }
}
```
Executing the query `isof(NS.DerivedCategory)` will throw the exception:
```
"Encountered invalid type cast. 'NS.Category' is not assignable from 'NS.Product'."
```

This occurs because ODL checks if `NS.Product` is related to `NS.Category` using the [CheckRelatedTo ](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Binders/DottedIdentifierBinder.cs#L104) function and since the types are not related or one is not a derivative of the other, the exception is thrown.

### Main changes
According to the [spec_cast](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_cast) and [spec_isof](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_isof) specifications, there is no indication that an exception should be thrown if casting fails. This pull request introduces the `TryBindDottedIdentifierForIsOfOrCastFunctionCall` method, which handles the binding of `DottedIdentifierToken` specifically for the `isof` and `cast` function calls without performing the `CheckRelatedTo` validation, thereby preventing exceptions when type casting is not possible.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

habbes commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -234,7 +234,22 @@ internal QueryNode BindFunctionCall(FunctionCallToken functionCallToken)
             // If there isn't, bind as Uri function
             // Bind all arguments
-            List<QueryNode> argumentNodes = new List<QueryNode>(functionCallToken.Arguments.Select(ar => this.bindMethod(ar)));
+            List<QueryNode> argumentNodes = new List<QueryNode>(functionCallToken.Arguments.Select(argument =>
```
It may be better to call `functionCallToken.Arguments.Select(...).ToList()` than `new List(...)`. If `Arguments` has a known collection size, then using `ToList()` could use an optimized path that avoids resizing. See: 
https://github.com/dotnet/runtime/issues/107560

habbes commented on test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/FilterAndOrderByFunctionalTests.cs:
```diff
@@ -844,27 +844,31 @@ public void IsOfFunctionWithTwoParameters_WithoutSingleQuotesOnTypeParameter_Sho
+        [InlineData("cast(MyAddress,Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        [InlineData("cast(null,Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        [InlineData("cast('',Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        public void CastFunctionWithUnquotedTypeParameter_WithIncorrectType_DoesNotThrowException(string filterQuery)
```
What is the expected behaviour of `cast()` when the types are incorrect? If it doesn't throw an exception, what will happen? How is this scenario expected to be handled?

WanjohiSammy replied:
This is what is happening currently with `cast` with `quoted type parameter`
- ODL will bind `literal type parameter token` to `ConstantNode` using [BindLiteral(LiteralToken literalToken)](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Binders/LiteralBinder.cs#L21). The AspNetCore library will get the `cast` node and tries to create LINQ Expression using [Expression BindCastSingleValue(SingleValueFunctionCallNode node, QueryBinderContext context)](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.SingleValueFunctionCall.cs#L562). It knows how to handle `ConstantNode`

For `cast` with `unquoted type parameter`
- ODL will bind `DottedIDentifierToken` as `SingleResourceCastNode` using a method I have added with this PR. To support this in AspNetCore, I will update function [Expression BindCastSingleValue(SingleValueFunctionCallNode node, QueryBinderContext context)](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.SingleValueFunctionCall.cs#L562) to add a condition to cast QueryNode to SingleResourceCastNode like what was done with the PR: https://github.com/OData/AspNetCoreOData/pull/1313

habbes replied:
Thanks for this added context. Can you also shed light on the behaviour with respect to handling of types where a cast is not possible? Ideally, from the user's endpoint the behaviour should be the same regardless of whether they user quoted or non-quoted syntax, and regardless of our internal implementation details which you have highlighted.

So could you clarify how we handle both supported and invalid cast scenarios in both quoted and unquoted variants? In which cases are we throwing an exception? Is the exception thrown consistent? Is that the expected behaviour?

My expectation is that `isof` should not throw an exception when a cast is invalid, but rather evaluate to a boolean that will perform the correct filter. I think `cast` throwing an exception when a cast is invalid is reasonable. What is not clear to me is whether such an exception should be thrown in ODL or AspNetCore. For the case of `cast` I would be cautious about changing the existing behaviour to avoid introducing a breaking change.

WanjohiSammy replied:
With the current behavior, ODL does not throw exception or to be more specific, it doesn't check if target type is related to source type. It only bind the `quoted type parameters` to `constant node` and return the entire token as FunctionCall token. The `AspNetCore`, on the other end, will get the token and try to create LINQ expression. If expression is not possible, AspNetCore returns null [BindSingleResourceCastFunctionCall](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.cs#L655) will return `null` expression. 

Let's look at this query:
```
/products?$filter=cast('Microsoft.AspNetCore.OData.E2E.Tests.Cast.Category')/Name eq 'Cat
```

AspNetCore will try to relate `Entity Product` to `Entity Category` in [BindSingleResourceCastFunctionCall](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.cs#L655) and returns null expression if one is not assignable from the other.

And then [CreatePropertyAccessExpression](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.cs#L1114) will throw an expression when trying to get the Property Name from a null expression.

![image](https://github.com/user-attachments/assets/ab57583f-99fc-43bc-9a6c-502cf206e2ae)

An exception like below will be thrown:
```
 Message: 
System.ArgumentException : Instance property 'Name' is not defined for type 'System.Object' (Parameter 'propertyName')

  Stack Trace: 
Expression.Property(Expression expression, String propertyName)
QueryBinder.CreatePropertyAccessExpression(Expression source, QueryBinderContext context, IEdmProperty property, String propertyPath) line 1124
QueryBinder.BindPropertyAccessQueryNode(SingleValuePropertyAccessNode propertyAccessNode, QueryBinderContext context) line 360
QueryBinder.BindSingleValueNode(SingleValueNode node, QueryBinderContext context) line 147
QueryBinder.Bind(QueryNode node, QueryBinderContext context) line 73
QueryBinder.BindBinaryOperatorNode(BinaryOperatorNode binaryOperatorNode, QueryBinderContext context) line 453
QueryBinder.BindSingleValueNode(SingleValueNode node, QueryBinderContext context) line 132
QueryBinder.Bind(QueryNode node, QueryBinderContext context) line 73
FilterBinder.BindFilter(FilterClause filterClause, QueryBinderContext context) line 39
```

habbes commented on test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/FilterAndOrderByFunctionalTests.cs:
```diff
@@ -844,27 +844,31 @@ public void IsOfFunctionWithTwoParameters_WithoutSingleQuotesOnTypeParameter_Sho
+        [InlineData("cast(MyAddress,Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        [InlineData("cast(null,Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        [InlineData("cast('',Fully.Qualified.Namespace.Employee)/WorkID eq 345")]
+        public void CastFunctionWithUnquotedTypeParameter_WithIncorrectType_DoesNotThrowException(string filterQuery)
```
Is this behaviour consistent with how cast is handled when the type param is quoted?

WanjohiSammy replied:
- For `quoted type parameter`, ODL bind the token to `ConstantNode` and AspNetCore is responsible to create LINQ expression using [Expression BindCastSingleValue(SingleValueFunctionCallNode node, QueryBinderContext context)](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.SingleValueFunctionCall.cs#L562). It knows how to handle `ConstantNode`.
- For `unquoted type parameter`, ODL bind the token to `SingleResourceCastNode`. When bumping the ODL version in AspNetCore, I will have to update the method [Expression BindCastSingleValue(SingleValueFunctionCallNode node, QueryBinderContext context)](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.SingleValueFunctionCall.cs#L562) to allow cast to `SingleResourceCastNode` as well.

habbes commented on test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/FilterAndOrderByFunctionalTests.cs:
```diff
@@ -844,27 +844,31 @@ public void IsOfFunctionWithTwoParameters_WithoutSingleQuotesOnTypeParameter_Sho
+        [InlineData("isof(MyAddress,Fully.Qualified.Namespace.Pet1)")]
+        [InlineData("isof(null,Fully.Qualified.Namespace.Person)")]
+        [InlineData("isof('',Fully.Qualified.Namespace.Person)")]
+        public void IsOfFunctionsWithUnquotedTypeParameter_WithIncorrectType_DoesNotThrowException(string filterQuery)
```
Is this behaviour consistent with how `isof` is handled when the type param is quoted?

WanjohiSammy replied:
Same applied to `isof` as discussed in https://github.com/OData/odata.net/pull/3117#discussion_r1840468666, however, the method in `AspNetCore` that is responsible to create LINQ is different. Here is the method [Expression BindIsOf(SingleValueFunctionCallNode node, QueryBinderContext context)](https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.SingleValueFunctionCall.cs#L637). I will also update this method to allow cast to `SingleResourceCastNode` as it currently only support cast to `ConstantNode`

marabooy commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -426,6 +436,52 @@ private bool TryBindIdentifier(string identifier, IEnumerable<FunctionParameterT
+            }
+
+            IEdmSchemaType childType = UriEdmHelpers.FindTypeFromModel(state.Model, dottedIdentifierToken.Identifier, this.Resolver);
+            IEdmStructuredType childStructuredType = childType as IEdmStructuredType;
+
+            if (childStructuredType == null)
+            {
+                return this.bindMethod(dottedIdentifierToken);
+            }
```
```suggestion

            if ( childType is IEdmStructuredType childStructuredType)
            {
                return this.bindMethod(dottedIdentifierToken);
            }

```
nit: Use is operator instead

WanjohiSammy replied:
Thanks for this. Implemented

xuzhg commented on test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/EnumFilterFunctionalTests.cs:
```diff
@@ -409,20 +409,24 @@ public void ParseFilterWithNullEnumValue()
+        [Theory]
+        [InlineData("cast(NS.Color'Green', 'Edm.String') eq 'blue'")]
+        [InlineData("cast(NS.Color'Green', Edm.String) eq 'blue'")]
+        public void ParseFilterCastMethod1(string filterQuery)
```
The test name contains XXX1 AND XXX2 doesn't make sense. Can you rename them?

And can combine 'ParseFilterCastMethod1' and 'ParseFilterCastMethod2'?

WanjohiSammy replied:
Updated

habbes commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -234,7 +234,27 @@ internal QueryNode BindFunctionCall(FunctionCallToken functionCallToken)
+            List<QueryNode> argumentNodes = functionCallToken.Arguments.Select(argument =>
+            {
+                // If the function is IsOf or Cast and the argument is a dotted identifier, we need to bind it differently
+                if (UnboundFunctionNames.Contains(functionCallToken.Name) && argument.ValueToken is DottedIdentifierToken dottedIdentifier)
```
Does `UnboundFunctionNames` check for `isof` and `cast`?

habbes commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -234,7 +234,27 @@ internal QueryNode BindFunctionCall(FunctionCallToken functionCallToken)
+                    IEdmSchemaType dottedIdentifierType = UriEdmHelpers.FindTypeFromModel(state.Model, dottedIdentifier.Identifier, this.Resolver);
+
+                    // If the dotted identifier is a primitive type and the next token is not null, we need to ensure next token is null
+                    if (dottedIdentifierType is IEdmPrimitiveType && dottedIdentifier.NextToken != null)
```
Why do we need to ensure that the next token is null?

WanjohiSammy replied:
Consider the following:
- `cast` or `isof` with 2 parameters. For example, cast(ID, Edm.String):
   - The parser will create a `FunctionCallToken` with 2 arguments where the first argument (`ID`) is an `EndPathToken` and the second argument (`Edm.String`) is a `DottedIdentifierToken` because it is unquoted. When the type parameter is quoted, the second argument becomes a `LiteralToken`.
   - The second argument's Next Token will point to the first argument. This is important for determining if the type of the second argument can be assigned from the type of the first argument.
   - For primitive DottedIdentifier types like Edm.String, Edm.Int32, Edm.Byte (including their lowercase variants), the presence of NextToken will cause the DottedIdentifier to bind as a `SingleValueCastNode`, leading to potential breaking changes as reported in issue https://github.com/OData/odata.net/issues/3123
   
- `cast` or `isof` with 1 parameter. For example, cast(Edm.String):
  - The parser will create a `FunctionCallToken` with 1 argument where the argument (`Edm.String`) is a `DottedIdentifierToken`. This argument does not have a NextToken.

Here we are checking DottedIdentifier is primitive type and next token is null to ensure we handle the primitive types properly and cases where argument is 1.

habbes commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -234,7 +234,27 @@ internal QueryNode BindFunctionCall(FunctionCallToken functionCallToken)
+                    if (dottedIdentifierType is IEdmPrimitiveType && dottedIdentifier.NextToken != null)
+                    {
+                        // Push the current argument onto the stack to keep track of the previous argument
+                        dottedIdentifier.NextToken = null;
```
Is it safe to modify tokens like this? Is this something we do elsewhere in the code?

WanjohiSammy replied:
Instead of `dottedIdentifier.NextToken = null` I have opted to do the following:
```cs
// If the next token is not null, we need to ensure it is null
dottedIdentifier = new DottedIdentifierToken(dottedIdentifier.Identifier, null);
```
This is done in [ParseIdentifier(QueryToken parent)](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Parsers/IdentifierTokenizer.cs#L71) where `parent` is the `NextToken`  and for DottedIdentifier, parent is always null

habbes commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -426,6 +446,50 @@ private bool TryBindIdentifier(string identifier, IEnumerable<FunctionParameterT
+        /// <exception cref="ODataException">Thrown when the token cannot be bound as a single resource node.</exception>
+        private QueryNode TryBindDottedIdentifierForIsOfOrCastFunctionCall(DottedIdentifierToken dottedIdentifierToken, IEdmSchemaType childType)
+        {
+            QueryNode parent = null;
```
Could you explain what is going on in this method?

xuzhg commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -234,7 +234,28 @@ internal QueryNode BindFunctionCall(FunctionCallToken functionCallToken)
+            List<QueryNode> argumentNodes = functionCallToken.Arguments.Select(argument =>
+            {
+                // If the function is IsOf or Cast and the argument is a dotted identifier, we need to bind it differently
+                string functionCallTokenName = IsUnboundFunction(functionCallToken.Name);
```
IsUnboundFunction returns a string? It looks non-sense.

WanjohiSammy replied:
`IsUnboundFunction` gets `functionCallToken.Name` in any case (lower-case, upper case or camel case) and compare with the `isof` and `cast` using `this.state.Configuration.EnableCaseInsensitiveUriFunctionIdentifier` config

[private string IsUnboundFunction(string functionName)](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs#L635)

[private static readonly string[] UnboundFunctionNames](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs#L29)

xuzhg commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -234,7 +234,28 @@ internal QueryNode BindFunctionCall(FunctionCallToken functionCallToken)
+                    // If the dotted identifier is a primitive type and the next token is not null, we need to ensure next token is null
+                    if (dottedIdentifierType is IEdmPrimitiveType && dottedIdentifier.NextToken != null)
+                    {
+                        // If the next token is not null, we need to ensure it is null
```
This comment line doesn't make sense?

gathogojr commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -758,7 +765,7 @@ private static IEdmTypeReference ValidateIsOfOrCast(BindingState state, bool isC
                 // throw if cast enum to not-string :
                 if ((args[0].GetEdmTypeReference() is IEdmEnumTypeReference)
-                    && !string.Equals(typeArgument.Value as string, Microsoft.OData.Metadata.EdmConstants.EdmStringTypeName, StringComparison.Ordinal))
+                    && !string.Equals(typeArgumentFullName, Microsoft.OData.Metadata.EdmConstants.EdmStringTypeName, state.Configuration.Resolver.EnableCaseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))
```
Can `state.Configuration.Resolver.` be null?

WanjohiSammy replied:
It can be null. In case the [DI container is nul](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs#L567)l, the Resolver will always be an instance of `ODataUriResolver()`:

https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs#L25

xuzhg commented on src/Microsoft.OData.Core/UriParser/Parsers/FunctionCallParser.cs:
```diff
@@ -33,6 +34,11 @@ internal sealed class FunctionCallParser : IFunctionCallParser
+        /// <summary>
+        /// If the function call is cast or isof, set to true.
+        /// </summary>
+        private bool isFunctionCallNameCastOrIsOf = false;
```
Why do you use the private to cache a transient state?
If it's a state/signal within the parser, please find a different way to save/cache it,?

WanjohiSammy replied:
Instead of having to cache, I have passed it as argument.

mikepizzo commented:
@WanjohiSammy -- I ran into this last week, and it may shortly be blocking some agentic API work.  What's the status of getting this fix in?

WanjohiSammy commented:
> I ran into this last week, and it may shortly be blocking some agentic API work. What's the status of getting this fix in?

@mikepizzo This is ready. I will remind the team to review.

mikepizzo commented:
/AzurePipelines run

gathogojr commented:
@WanjohiSammy I had a question regarding this part of your pull request description:

> According to the https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_cast and https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_isof specifications, there is no indication that an exception should be thrown if casting fails. This pull request introduces the `TryBindDottedIdentifierForIsOfOrCastFunctionCall` method, which handles the binding of `DottedIdentifierToken` specifically for the `isof` and `cast` function calls without performing the `CheckRelatedTo` validation, thereby preventing exceptions when type casting is not possible.

I understand the intent to align with the spec and avoid throwing exceptions on failed casts. However, I’m wondering about the practical implications of this behavior.

For example, consider the following scenario:

```csharp
public class ProductsController : ODataController
{
    [EnableQuery]
    public ActionResult<IEnumerable<Product>> Get()
    {
        return new List<Product>
        {
            new Product { Id = 1, Name = "Banana" },
            new Product { Id = 2, Name = "Orange" },
            new Product { Id = 3, Name = "Apple" }
        };
    }
}
```

With the following model:

```csharp
namespace NS
{
    public class Product
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }

    public class DerivedProduct : Product
    {
        public DateTimeOffset ExpiryDate { get; set; }
    }

    public class Category
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }
}
```

Now, if someone runs:

```
/Products?$filter=isof(NS.DerivedProduct)
/Products?$filter=isof(NS.Category)
```

Both filters would yield the same result—an empty set—because there are no `DerivedProduct` instances in the collection, and `Category` is unrelated to `Product`.

While this behavior is technically correct per the spec, it raises a usability concern: if a user mistakenly writes `isof(NS.Category)` thinking it's a valid filter, the system silently returns no results. This could lead to confusion or debugging difficulties, especially when the type is completely unrelated.

I understand the idea of staying spec-compliant but how can we help users catch likely mistakes when the cast target is unrelated to the base type?

Curious to hear your thoughts on this trade-off.
cc @mikepizzo

gathogojr commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -720,12 +720,19 @@ private static IEdmTypeReference ValidateIsOfOrCast(BindingState state, bool isC
             }
 
-            ConstantNode typeArgument = args.Last() as ConstantNode;
+            QueryNode queryNode = args.Last();
```
Given the `args.Count` above this line, that suggests to me that we're dealing with a `List<T>`. I think we can avoid using LINQ here

gathogojr commented on src/Microsoft.OData.Core/UriParser/Parsers/FunctionCallParser.cs:
```diff
@@ -161,33 +163,54 @@ public FunctionParameterToken[] ParseArgumentListOrEntityKeyList(Action restoreA
         {
+            // Store the parent expression of the current argument.
+            Stack<QueryToken> expressionParents = new Stack<QueryToken>();
+            bool isFunctionCallNameCastOrIsOf = functionName.Length > 0 &&
+                functionName.SequenceEqual(ExpressionConstants.UnboundFunctionCast.AsSpan()) || functionName.SequenceEqual(ExpressionConstants.UnboundFunctionIsOf.AsSpan());
+
```
@WanjohiSammy Are you going to consider this suggestion by Copilot?

gathogojr commented on src/Microsoft.OData.Core/UriParser/Parsers/FunctionCallParser.cs:
```diff
@@ -213,5 +236,37 @@ private bool TryReadArgumentsAsNamedValues(out ICollection<FunctionParameterToke
+            if (parameterToken is DottedIdentifierToken dottedIdentifierToken && dottedIdentifierToken.NextToken == null)
+            {
+                // Check if the dottedIdentifier is a primitive type
+                IEdmSchemaElement schemaElement = EdmCoreModel.Instance.SchemaElements.Where(e => string.Equals(dottedIdentifierToken.Identifier, e.FullName(), StringComparison.OrdinalIgnoreCase)).FirstOrDefault();
```
Do we really need to loop through all the schema elements in the EdmCoreModel to confirm if `dottedIdentifierToken.Identifier` is primitive? I think in ODL there are checks for primitive that don't involve scanning through the EdmCoreModel schema elements using LINQ

gathogojr replied:
Just realized that identifier will look something like this:
<img width="441" height="113" alt="image" src="https://github.com/user-attachments/assets/2f343bb2-79ae-4f13-81ed-24f118954619" />

All the same, I think we should find a more efficient way of evaluating that. Maybe we can maintain an unordered set (e.g. `new HashSet<string>(StringComparer.OrdinalIgnoreCase);`) of primitive identifiers (in a class variable) that we initialize at first use and subsequently reuse it for similar checks. Use of EdmCoreModel.Instance.SchemaElements feel a bit too suboptimal for me

WanjohiSammy replied:
@habbes The real scenario is that we will need to initialize `NormalizedModelElementsCache` or `ODataUriResolver`, that is why the initialization is not outside the benchmark.

WanjohiSammy replied:
Added a `singleton` for `EdmCoreModel` in `NormalizedModelElementsCache` has significantly reduce the 

<img width="1399" height="569" alt="image" src="https://github.com/user-attachments/assets/d7f5c279-0f14-4b8f-bb66-716f2f9dd335" />

### Note
- Using of `ODataResolver` still performs poorly even if the initialization is done outside the benchmark.
- `EdmCoreModelInstance` is a singleton specifically for `EDM core model`
   ```cs
   public static NormalizedModelElementsCache EdmCoreModelCacheInstance = new NormalizedModelElementsCache(EdmCoreModel.Instance);
  ```

### Benchmarks code
```cs
[MemoryDiagnoser]
[CPUUsageDiagnoser]
public class MyBenchmarks
{
    [Params("", "edm.int64", "edm.string", "Edm.String", "EDM.decimal", "EDM.GUID", "fully.qualified.ns")]
    public string TypeName { get; set; }

    private ODataUriResolver Resolver { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        Resolver = new ODataUriResolver { EnableCaseInsensitive = true };
    }

    [Benchmark(Baseline = true)]
    public object? Using_EdmCoreModel_CoreSchemaElements()
    {
        return EdmCoreModel.Instance.SchemaElements.FirstOrDefault(se => string.Equals(se.Namespace, TypeName, StringComparison.OrdinalIgnoreCase));
    }

    [Benchmark]
    public object? Using_NormalizedModelElementsCache_FindSchemaTypes()
    {
        var schemaTypes = NormalizedModelElementsCache.EdmCoreModelInstance.FindSchemaTypes(TypeName);
        return schemaTypes?.FirstOrDefault();
    }

    [Benchmark]
    public object? Using_ODataUriResolver()
    {
        var edmSchemaType = Resolver.ResolveType(EdmCoreModel.Instance, TypeName);
        return edmSchemaType;
    }
}
```

habbes replied:
Interesting, the speedup is even more significant than I expected. I also find it curious that the ODataUriResolver is faster only when the type name is Edm.String.

gathogojr commented on src/Microsoft.OData.Core/UriParser/Parsers/FunctionCallParser.cs:
```diff
@@ -213,5 +236,37 @@ private bool TryReadArgumentsAsNamedValues(out ICollection<FunctionParameterToke
+        /// <param name="parentExpression">The previous parameter token.</param>
+        /// <param name="parameterToken">The current parameter token.</param>
+        /// <returns>The updated parameter token.</returns>
+        private static QueryToken SetParentForCurrentParameterToken(QueryToken parentExpression, QueryToken parameterToken)
+        {
+            // If the parameter is a dotted identifier, we need to set the parent of the next argument to the current argument.
+            // But if it is primitive literal, no need to set the parent.
+            if (parameterToken is DottedIdentifierToken dottedIdentifierToken && dottedIdentifierToken.NextToken == null)
+            {
+                // Check if the dottedIdentifier is a primitive type
+                IEdmSchemaElement schemaElement = EdmCoreModel.Instance.SchemaElements.Where(e => string.Equals(dottedIdentifierToken.Identifier, e.FullName(), StringComparison.OrdinalIgnoreCase)).FirstOrDefault();
+
+                // If the dottedIdentifier is not a primitive type, set the parent of the next argument to the current argument.
+                //  cast(1, Edm.Int32) -> Edm.Int32 is a dottedIdentifierToken but a primitive type
+                //  isof(1, edm.int64) -> edm.int64 is a dottedIdentifierToken but a primitive type
+                //  isof(MyAddress, NS.Address) -> NS.Address is a dottedIdentifierToken but not a primitive type
+                if (schemaElement is null or not IEdmPrimitiveType)
+                {
+                    dottedIdentifierToken.NextToken = parentExpression;
+                    parameterToken = dottedIdentifierToken;
+                }
+            }
+
+            return parameterToken;
+        }
```
```csharp
private static QueryToken SetParentForCurrentParameterToken(QueryToken parentExpression, QueryToken parameterToken)
{
    // If the parameter is not a DottedIdentifierToken or already has a NextToken
    if (parameterToken is not DottedIdentifierToken dottedIdentifierToken || dottedIdentifierToken.NextToken is not null)
    {
        return parameterToken;
    }

    // Check if the identifier is a primitive type
    IEdmSchemaElement schemaElement = EdmCoreModel.Instance.SchemaElements
        .FirstOrDefault(e => string.Equals(dottedIdentifierToken.Identifier, e.FullName(), StringComparison.OrdinalIgnoreCase));

    // If the identifier is a primitive type
    if (schemaElement is IEdmPrimitiveType)
    {
        return parameterToken;
    }

    // Set the parent of the next argument to the current argument
    dottedIdentifierToken.NextToken = parentExpression;
    return dottedIdentifierToken;
}
```

WanjohiSammy replied:
Updated

WanjohiSammy commented:
/AzurePipelines run

gathogojr commented on src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs:
```diff
@@ -720,12 +720,27 @@ private static IEdmTypeReference ValidateIsOfOrCast(BindingState state, bool isC
             }
 
-            ConstantNode typeArgument = args.Last() as ConstantNode;
+            QueryNode queryNode;
+            if (args.Count == 1)
+            {
+                queryNode = args[0];
+            }
+            else
+            {
+                queryNode = args[1];
+            }
```
@WanjohiSammy `args.Last()` would be equivalent to `args[args.Count - 1]`, no? Or is there a chance or having more than 2 elements in the args collection such that we want to particularly target the element at index 1 in all other cases where there's more than 1 elements?

WanjohiSammy replied:
Since `isof` and `cast` can have 1 or 2 parameters where the last parameter is the type, we can use your suggestion `args[args.Count - 1]` or `args[^1]` using the ["hat" operator (^)](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges#overview)

WanjohiSammy commented:
> Both filters would yield the same result—an empty set—because there are no `DerivedProduct` instances in the collection, and `Category` is unrelated to `Product`.
> 
> While this behavior is technically correct per the spec, it raises a usability concern: if a user mistakenly writes `isof(NS.Category)` thinking it's a valid filter, the system silently returns no results. This could lead to confusion or debugging difficulties, especially when the type is completely unrelated.
> 
> I understand the idea of staying spec-compliant but how can we help users catch likely mistakes when the cast target is unrelated to the base type?
> 
> Curious to hear your thoughts on this trade-off. cc @mikepizzo

@gathogojr I have updated the PR description of what the change is about. I have also added tests to ensure that exceptions are thrown when the types do not match.

habbes commented:
@WanjohiSammy the issues listed in the description that this PR aims to fix happen to be closed. What is the exact behaviour that this PR aims to change or rectify? Is it to support case-insensitivity in the `isof` and `cast` operator or supporting `isof` on unrelated types without throwing an error? It's not clear to me from reading the description what did not work before vs what now works after this PR.

habbes commented on src/Microsoft.OData.Core/UriParser/Parsers/FunctionCallParser.cs:
```diff
@@ -161,33 +163,53 @@ public FunctionParameterToken[] ParseArgumentListOrEntityKeyList(Action restoreA
+        private List<FunctionParameterToken> ReadArgumentsAsPositionalValues(ReadOnlySpan<char> functionName = default)
         {
+            // Store the parent expression of the current argument.
+            Stack<QueryToken> expressionParents = new Stack<QueryToken>();
```
nit: Since there are likely to be few arguments in most cases, and this stack is not used outside of this method, might be an option to consider a stack allocated array or a struct-based stack that only goes to the heap once it reaches a certain size threshold. Might be overkill for this PR, but I think there a couple of places where we build this short-lived single-scoped stacks that may end up causing avoidable allocations in the hot path. We can talk more about this if interested.

habbes replied:
Here's a sample implementation (I asked Copilot to generate the code)

```c#

using System;
using System.Collections.Generic;

using System.Runtime.CompilerServices;

[InlineArray(4)]
public struct InlineBuffer<T>
{
    private T _element0; // Compiler generates _element1, _element2, _element3
}

public struct StructStack<T>
{
    private InlineBuffer<T> _inline;
    private List<T>? _overflow;
    private int _count;

    private const int InlineCapacity = 4;

    public int Count => _count;

    public void Push(T item)
    {
        if (_count < InlineCapacity)
        {
            _inline[_count] = item;
        }
        else
        {
            _overflow ??= new List<T>();
            _overflow.Add(item);
        }

        _count++;
    }

    public T Pop()
    {
        if (_count == 0)
            throw new InvalidOperationException("Stack is empty.");

        _count--;

        if (_count < InlineCapacity)
        {
            return _inline[_count];
        }
        else
        {
            var item = _overflow![^1];
            _overflow.RemoveAt(_overflow.Count - 1);
            return item;
        }
    }

    public T Peek()
    {
        if (_count == 0)
            throw new InvalidOperationException("Stack is empty.");

        if (_count <= InlineCapacity)
        {
            return _inline[_count - 1];
        }
        else
        {
            return _overflow![_overflow.Count - 1];
        }
    }
}
```

WanjohiSammy replied:
I have something similar. I will push and let you review.

WanjohiSammy replied:
@habbes I have added the [StackStruct](https://github.com/OData/odata.net/pull/3117/files#diff-bf1d6cd4977105df5d6550120732acf99891f84b62b1b6f20ad9dfa10e36711e)

habbes commented on src/Microsoft.OData.Core/UriParser/Parsers/FunctionCallParser.cs:
```diff
@@ -213,5 +235,36 @@ private bool TryReadArgumentsAsNamedValues(out ICollection<FunctionParameterToke
+            }
+
+            // Check if the identifier is a primitive type
+            IEdmSchemaElement schemaElement = EdmCoreModel.Instance.SchemaElements
```
Is there where it's been suggested to do a cache look-up? How many elements do we have in the core model schema elements?

habbes commented on src/Microsoft.OData.Core/UriParser/Resolver/NormalizedModelElementsCache.cs:
```diff
@@ -20,6 +20,11 @@ namespace Microsoft.OData.Edm
+        /// <summary>
+        /// Provides a shared, normalized case-insensitive cache of model elements for the EDM core model.
+        /// </summary>
+        public static NormalizedModelElementsCache EdmCoreModelCacheInstance = new NormalizedModelElementsCache(EdmCoreModel.Instance);
```
Make it readonly.

```suggestion
        public static readonly NormalizedModelElementsCache EdmCoreModelCacheInstance = new NormalizedModelElementsCache(EdmCoreModel.Instance);

```

nit: since this is already in a class called cache, we could simplify the model. But feel free to ignore this part

```suggestion
        public static readonly NormalizedModelElementsCache EdmCoreModelInstance = new NormalizedModelElementsCache(EdmCoreModel.Instance);
```

or 

```suggestion
        public static NormalizedModelElementsCache EdmCoreModelCache = new NormalizedModelElementsCache(EdmCoreModel.Instance);

```

WanjohiSammy commented on src/Microsoft.OData.Core/StackStructOfT.cs:
```diff
@@ -0,0 +1,175 @@
+    /// </summary>
+    /// <param name="other">The stack to compare with the current stack.</param>
+    /// <returns>true if the stacks are equal; otherwise, false.</returns>
+    public readonly bool Equals(StackStruct<T> other)
```
I have reverted to use the regular `stack`

WanjohiSammy commented on src/Microsoft.OData.Core/StackStructOfT.cs:
```diff
@@ -0,0 +1,175 @@
+/// Provides efficient push, pop, and peek operations with dynamic resizing.
+/// </summary>
+/// <typeparam name="T">The type of elements in the stack.</typeparam>
+internal struct StackStruct<T>
```
@habbes I have reverted to use the regular `stack`
